Source/Error.swift

```diff
@@ -24,41 +24,428 @@
+        case failedToCreateInputStream(from: URL)
+    }
+    
+    /// `OutputStreamFailureReason` is the reason returned for an `AFError.multipartEncodingFailed` error.
+    ///
+    /// - writeFailed:          <#writeFailed description#>
+    /// - failedToCreateStream: <#failedToCreateStream description#>
+    /// - fileAlreadyExists:    <#fileAlreadyExists description#>
+    /// - invalidFile:          <#invalidFile description#>
+    public enum OutputStreamFailureReason {
+        case writeFailed(to: OutputStream)
+        case failedToCreateStream(at: URL)
+        case fileAlreadyExists(at: URL)
+        case invalidFile(at: URL)
+    }
+    
+    public enum SerializationFailureReason {
+        case inputDataNil
+        case inputDataNilOrZeroLength
+        case invalidStringEncoding(String.Encoding)
+    }
+    
+    public enum ValidationFailureReason {
+        case contentTypeMismatch(acceptable: [String], actual: String)
+        case contentTypeMissingAndNotWildcard
+        case unacceptableStatusCode(code: Int)
+    }
+    
+    case multipartEncodingFailed(reason: MultipartEncodingFailureReason)
+    case inputStreamReadFailed(inputStream: InputStream)
+    case outputStreamWriteFailed(reason: OutputStreamFailureReason)
+    case responseValidationFailed(reason: ValidationFailureReason)
+    case responseSerializationFailed(reason: SerializationFailureReason)
+}
+
+// MARK: Error Booleans
+
+public extension AFError {
+    
+    // MARK: Multipart Encoding
+    
+    /// Returns whether or not the AFError is a multipart encoding error. When this property is true,
+    /// the `url` property will contain the associated `URL`.
+    var isMultipartEncodingError: Bool {
+        if case .multipartEncodingFailed = self { return true }
+        return false
+    }
+    
+    var isMultipartNotAFileError: Bool {
+        guard case let .multipartEncodingFailed(reason) = self else { return false }
+        return reason.isNotAFile
+    }
+    
+    var isMultipartFailedToExtractFilenameError: Bool {
+        guard case let .multipartEncodingFailed(reason) = self else { return false }
+        return reason.isFailedToExtractFilename
+    }
+    
+    var isMultipartFileNotReachableError: Bool {
+        guard case let .multipartEncodingFailed(reason) = self else { return false }
+        return reason.isFileNotReachable
+    }
+    
+    var isMultipartIsADirectoryError: Bool {
+        guard case let .multipartEncodingFailed(reason) = self else { return false }
+        return reason.isADirectory
+    }
+    
+    var isMultipartFailedToFetchAttributesError: Bool {
+        guard case let .multipartEncodingFailed(reason) = self else { return false }
+        return reason.isFailedToFetchAttributes
+    }
+    
+    var isMultipartFailedToCreateInputStreamError: Bool {
+        guard case let .multipartEncodingFailed(reason) = self else { return false }
+        return reason.isFailedToCreateInputStream
+    }
```

Comment:
I'm on the fence with these properties. I like the ones for the top level cases, but I question creating properties for every `XXXFailureReason`. It seems a bit excessive. Are we really thinking that people won't want to use switches? These are just so verbose it takes a while to figure out what one you'd even want to call.